[Docs] PITR: Changes documenting PITR support for YSQL DDL (#9570)
yugabyte/yugabyte-db: docs/content/latest/explore/cluster-management/point-in-time-recovery-ysql.md

  </li>
</ul>

> The point-in-time recovery feature allows you to restore the state of your cluster's data (and some types of metadata) from a specific point in time. This can be relative, such as "three hours ago", or an absolute timestamp.

Reviewer: ```suggestion
The point-in-time recovery feature allows you to restore the state of your cluster's data (and [certain types](../../../manage/backup-restore/point-in-time-recovery/#limitations) of metadata) from a specific point in time. This can be relative, such as "three hours ago", or an absolute timestamp.
```

**Careful!** If you specify a time prior to when you created the table, the restore will leave the table intact, but empty.

Refer to the yb-admin [_restore-snapshot-schedule_ command](../../../admin/yb-admin/#restore-snapshot-schedule) for more details.

## Undo metadata changes

In addition to data changes, you can also use PITR to recover from metadata changes, such as creating, altering, and deleting tables and indexes.

### Undo table creation

> 1. Create a snapshot schedule for the database from a shell prompt. In this example, the schedule is on the default database yugabyte, one snapshot every minute, and each snapshot is retained for ten minutes.

Reviewer: ```suggestion
1. At a terminal prompt, create a snapshot schedule for the database. In this example, the schedule is on the default `yugabyte` database, one snapshot every minute, and each snapshot is retained for ten minutes.
```

    ```sh
    $ bin/yb-admin create_snapshot_schedule 1 10 ysql.yugabyte
    ```

    ```output
    {
      "schedule_id": "1ccb7e8b-4032-48b9-ac94-9f425d270a97"
    }
    ```

1. Verify that a snapshot has happened.

    ```sh
    $ bin/yb-admin list_snapshot_schedules
    ```

    ```output
    {
      "schedules": [
          {
              "id": "1ccb7e8b-4032-48b9-ac94-9f425d270a97",
              "options": {
                  "filter": "ysql.yugabyte",
                  "interval": "1 min",
                  "retention": "10 min"
              },
              "snapshots": [
                  {
                      "id": "94052190-1f39-44f3-b66f-87e40e1eca04",
                      "snapshot_time": "2021-08-02 22:22:55.251562"
                  }
              ]
          }
      ]
    }
    ```

> 1. In order to restore from an absolute time, get a timestamp from the command prompt. We'll create a table and then restore to this time in order to undo the table creation.

Reviewer: ```suggestion
1. To restore from an absolute time, get a timestamp from the command prompt. You'll create a table, then restore to this time to undo the table creation.
```

      VALUES 
      (1221, 'John Smith', 'Marketing', 50000),
      (1222, 'Bette Davis', 'Sales', 55000),
      (1223, 'Lucille Ball', 'Operations', 70000),
      (1224, 'John Zimmerman', 'Sales', 60000);
    
    SELECT * from employees;
    ```

    ```output
     employee_no |      name      | department | salary 
    -------------+----------------+------------+--------
            1223 | Lucille Ball   | Operations |  70000
            1224 | John Zimmerman | Sales      |  60000
            1221 | John Smith     | Marketing  |  50000
            1222 | Bette Davis    | Sales      |  55000
    (4 rows)
    ```

1. Restore the snapshot schedule to the timestamp you obtained before you created the table, at a terminal prompt.

    ```sh
    $ bin/yb-admin restore_snapshot_schedule 1ccb7e8b-4032-48b9-ac94-9f425d270a97 1627943076717734
    ```

    ```output
    {
      "snapshot_id": "5911ba63-9bde-4170-917e-2ee06a686e12",
      "restoration_id": "e059741e-1cff-4cf7-99c5-3c351c0ce22b"
    }
    ```

1. Next, verify the restoration is in `RESTORED` state (you'll see more snapshots in the list, as well).

    ```sh
    $ bin/yb-admin list_snapshots
    ```

    ```output

>     SSnapshot UUID                    	State

Reviewer: Also, replace ALL HARD TABS in the file with the correct number of spaces.

    94052190-1f39-44f3-b66f-87e40e1eca04 	COMPLETE
    d4e9879d-1873-4533-9a09-c0cd1aa34317 	COMPLETE
    5911ba63-9bde-4170-917e-2ee06a686e12 	COMPLETE
    05ae7198-a2a3-4374-b4ee-49ba38c8bc74 	COMPLETE
    6926fdd4-7cec-408e-b247-511b504499c1 	COMPLETE
    eec02516-c10a-4369-8ff5-ed1c7f129749 	COMPLETE
    Restoration UUID                 	  State
    e059741e-1cff-4cf7-99c5-3c351c0ce22b 	RESTORED
    ```

1. Verify that the table no longer exists.

    ```sh
    $ bin/ysqlsh -d yugabyte;
    $ \d employees;
    ```

### Undo table deletion

> 1. Create a snapshot schedule for the database from a shell prompt. In this example, the schedule is on the default database yugabyte, one snapshot every minute, and each snapshot is retained for ten minutes.

Reviewer: ```suggestion
1. From a terminal prompt, create a snapshot schedule for the database. In this example, the schedule is on the default `yugabyte` database, one snapshot every minute, and each snapshot is retained for ten minutes.
```

    (4 rows)
    ```

1. Verify that a snapshot has happened since table creation.

    ```sh
    $ bin/yb-admin list_snapshot_schedules
    ```

    ```output
      {
      "schedules": [
          {
              "id": "b4217ea5-56dc-4daf-afea-743460ece241",
              "options": {
                  "filter": "ysql.yugabyte",
                  "interval": "1 min",
                  "retention": "10 min"
              },
              "snapshots": [
                  {
                      "id": "2739695f-7b61-4996-98b3-c2a4052fd840",
                      "snapshot_time": "2021-08-03 11:24:28.632182"
                  },
                  {
                      "id": "0894192c-6326-4110-a5c3-fbdaaaac7d98",
                      "snapshot_time": "2021-08-03 11:25:33.641747",
                      "previous_snapshot_time": "2021-08-03 11:24:28.632182"
                  },
                  {
                      "id": "17364e01-e0e3-4ec5-a0e7-d69c45622351",
                      "snapshot_time": "2021-08-03 11:26:38.652024",
                      "previous_snapshot_time": "2021-08-03 11:25:33.641747"
                  }
              ]
          }
      ]
      }
    ```

> 1. In order to restore from an absolute time, get a timestamp from the command prompt. We'll delete the table and then restore to this time in order to undo the delete.

Reviewer: ```suggestion
1. To restore from an absolute time, get a timestamp from the command prompt. You'll delete the table, then restore to this time to undo the deletion.
```

    ```sh
    $ python -c 'import datetime; print datetime.datetime.now().strftime("%s%f")'
    ```

    ```output
    1627990118725202
    ```

1. Drop this table.

    ```sh
    $ drop table employees;
    ```

1. Restore the snapshot schedule to the timestamp you obtained before you deleted the table, at a terminal prompt.

    ```sh
    $ bin/yb-admin restore_snapshot_schedule b4217ea5-56dc-4daf-afea-743460ece241 1627990118725202
    ```

    ```output
    {
      "snapshot_id": "663cec5d-48e7-4f27-89ac-94c2dd0a3c32",
      "restoration_id": "eda28aa5-10bc-431d-ade9-44c9b8d1810e"
    }
    ```

1. Next, verify the restoration is in `RESTORED` state (you'll see more snapshots in the list, as well).

    ```sh
    $ bin/yb-admin list_snapshots
    ```

    ```output

>     Snapshot UUID                    	State

Reviewer: More hard tabs here, probably elsewhere too

              "options": {
                  "filter": "ysql.yugabyte",
                  "interval": "1 min",
                  "retention": "10 min"
              },
              "snapshots": [
                  {
                      "id": "ded348d6-a046-4778-8574-edb793739c37",
                      "snapshot_time": "2021-08-03 11:59:20.979156"
                  },
                  {
                      "id": "e2bfb948-8f24-4a0a-877f-a792ee9c969d",
                      "snapshot_time": "2021-08-03 12:00:25.988691",
                      "previous_snapshot_time": "2021-08-03 11:59:20.979156"
                  },
                  {
                      "id": "93a58196-d758-4bc8-93f7-72db8d334864",
                      "snapshot_time": "2021-08-03 12:01:30.999535",
                      "previous_snapshot_time": "2021-08-03 12:00:25.988691"
                  },
                  {
                      "id": "3f29453c-1159-4199-b2b5-558ebf14702f",
                      "snapshot_time": "2021-08-03 12:02:36.009326",
                      "previous_snapshot_time": "2021-08-03 12:01:30.999535"
                  }
              ]
          }
      ]
      }
    ```

> 1. In order to restore from an absolute time, get a timestamp from the command prompt. We'll add a column to the table and then restore to this time in order to undo the added column.

Reviewer: ```suggestion
1. To restore from an absolute time, get a timestamp from the command prompt. You'll add a column to the table, then restore to this time to undo the column addition.
```

    ```sh
    $ python -c 'import datetime; print datetime.datetime.now().strftime("%s%f")'
    ```

    ```output
    1627992256752809
    ```

1. Using the same database, alter your table by adding a column.

    ```sql
    alter table employees add column v2 int;
    select * from employees;
    ```

    ```output
     employee_no |      name      | department | salary | v2 
    -------------+----------------+------------+--------+----
            1223 | Lucille Ball   | Operations |  70000 |   
            1224 | John Zimmerman | Sales      |  60000 |   
            1221 | John Smith     | Marketing  |  50000 |   
            1222 | Bette Davis    | Sales      |  55000 |   
    (4 rows)
    ```

> 1. Restore the snapshot schedule to the timestamp you obtained before you added the column, at a terminal prompt.

Reviewer: ```suggestion
1. At a terminal prompt, restore the snapshot schedule to the timestamp you obtained before you added the column.
```

    ```sh
    $ bin/yb-admin restore_snapshot_schedule 47fd40c3-1c2f-4e1b-b64b-6c2c9f698946 1627992256752809
    ```

    ```output
    {
      "snapshot_id": "db876700-d553-49e4-a0d1-4c88e52d8a78",
      "restoration_id": "c240d26c-cbeb-46eb-b9ea-0a3e6a734ecf"
    }
    ```

1. Next, verify the restoration is in `RESTORED` state (you'll see more snapshots in the list, as well):

    ```sh
    $ bin/yb-admin list_snapshots
    ```

    ```output

>     Snapshot UUID                    	State

Reviewer: More hard tabs...

    (4 rows)
    ```

1. Verify that a snapshot has happened since table creation.

    ```sh
    $ bin/yb-admin list_snapshot_schedules
    ```

    ```output
    {
      "schedules": [
          {
              "id": "064d1734-377c-4842-a95e-88ce68c93ca9",
              "options": {
                  "filter": "ysql.yugabyte",
                  "interval": "1 min",
                  "retention": "10 min"
              },
              "snapshots": [
                  {
                      "id": "967843c0-58b6-4bd1-943a-5a75a6d2588d",
                      "snapshot_time": "2021-08-03 12:17:47.471817"
                  },
                  {
                      "id": "6b77a0a2-1ecc-4d5d-8e13-4bb7b57cca1e",
                      "snapshot_time": "2021-08-03 12:18:52.482001",
                      "previous_snapshot_time": "2021-08-03 12:17:47.471817"
                  },
                  {
                      "id": "99ab0e01-af59-4a6a-8870-9de6bd535f6d",
                      "snapshot_time": "2021-08-03 12:19:57.492521",
                      "previous_snapshot_time": "2021-08-03 12:18:52.482001"
                  }
              ]
          }
      ]
    }
    ```

> 1. In order to restore from an absolute time, get a timestamp from the command prompt. We'll remove a column from the table and then restore to this time in order to get the column back.

Reviewer: ```suggestion
1. To restore from an absolute time, get a timestamp from the command prompt. You'll remove a column from the table, then restore to this time to get the column back.
```

      "restoration_id": "3601225d-21ff-45e8-bebc-ff84c058d290"
    }
    ```

1. Next, verify the restoration is in `RESTORED` state (you'll see more snapshots in the list, as well):

    ```sh
    $ bin/yb-admin list_snapshots
    ```

    ```output
    Snapshot UUID                    	State
    967843c0-58b6-4bd1-943a-5a75a6d2588d 	COMPLETE
    6b77a0a2-1ecc-4d5d-8e13-4bb7b57cca1e 	COMPLETE
    99ab0e01-af59-4a6a-8870-9de6bd535f6d 	COMPLETE
    e798b0c5-607c-4662-8b70-beb7ef672ab6 	COMPLETE
    814982b2-2d86-425a-8ea1-29e356e5de1f 	COMPLETE
    980a86e3-88f3-4ed9-815f-6ff7923d2bff 	COMPLETE
    Restoration UUID                 	State
    3601225d-21ff-45e8-bebc-ff84c058d290 	RESTORED
    ```

1. Verify that the salary column is back.

    ```sql
    select * from employees;
    ```

    ```output
     employee_no | name           | department | salary
    -------------+----------------+------------+--------
            1223 |   Lucille Ball | Operations |  70000
            1224 | John Zimmerman |      Sales |  60000
            1221 |     John Smith |  Marketing |  50000
            1222 |    Bette Davis |      Sales |  55000

    (4 rows)
    ```

### Undo index creation

> 1. Create a snapshot schedule for the database from a shell prompt. In this example, the schedule is on the default database yugabyte, one snapshot every minute, and each snapshot is retained for ten minutes.

Reviewer: ```suggestion
1. From a terminal prompt, create a snapshot schedule for the database. In this example, the schedule is on the default `yugabyte` database, one snapshot every minute, and each snapshot is retained for ten minutes.
```

    (4 rows)
    ```

1. Verify that a snapshot has happened since table creation.

    ```sh
    $ bin/yb-admin list_snapshot_schedules
    ```

    ```output
    {
      "schedules": [
          {
              "id": "dcbe46e3-8108-4d50-8601-423b27d230b1",
              "options": {
                  "filter": "ysql.yugabyte",
                  "interval": "1 min",
                  "retention": "10 min"
              },
              "snapshots": [
                  {
                      "id": "b4316e4e-a7c2-49a3-af16-d928aef5630e",
                      "snapshot_time": "2021-08-03 12:36:56.212541"
                  },
                  {
                      "id": "c318b620-490c-4294-b495-a7f0349017d0",
                      "snapshot_time": "2021-08-03 12:38:01.221749",
                      "previous_snapshot_time": "2021-08-03 12:36:56.212541"
                  },
                  {
                      "id": "85dc006f-ebf5-464e-81f7-a406a3322942",
                      "snapshot_time": "2021-08-03 12:39:06.232231",
                      "previous_snapshot_time": "2021-08-03 12:38:01.221749"
                  }
              ]
          }
      ]
    }
    ```

> 1. In order to restore from an absolute time, get a timestamp from the command prompt. We'll create an index on the table and then restore to this time in order to undo the index creation.

Reviewer: ```suggestion
1. To restore from an absolute time, get a timestamp from the command prompt. You'll create an index on the table, then restore to this time to undo the index creation.
```
